File: src/storage/batchWriter.ts
import { Embedder } from "./embedder";
import { ElasticDB, MEMORIES_INDEX, SKILLS_INDEX, CONTEXT_INDEX } from "./elasticDB";

export interface BatchWriterOptions {
  batchSize?: number;
  flushIntervalMs?: number;
}

export interface BatchOp {
  type: "memory" | "skill" | "node";
  data: Record<string, any>;
}

export interface BatchResult {
  successful: number;
  failed: number;
  errors: Array<{ id: string; error: string }>;
}

const INDEX_MAP: Record<string, string> = {
  memory: MEMORIES_INDEX,
  skill: SKILLS_INDEX,
  node: CONTEXT_INDEX,
};

function getEmbedText(op: BatchOp): string {
  switch (op.type) {
    case "memory":
      return op.data.content;
    case "skill":
      return `${op.data.name}: ${op.data.description}`;
    case "node":
      return `${op.data.name}: ${op.data.abstract}`;
  }
}

function getId(op: BatchOp): string {
  return op.type === "node" ? op.data.uri : op.data.id;
}

export class BatchWriter {
  private db: ElasticDB;
  private queue: BatchOp[] = [];
  private readonly batchSize: number;
  private flushTimer: ReturnType<typeof setInterval> | null = null;
  private readonly flushIntervalMs: number;

  constructor(db: ElasticDB, options: BatchWriterOptions = {}) {
    this.db = db;
    this.batchSize = options.batchSize ?? 10;
    this.flushIntervalMs = options.flushIntervalMs ?? 2000;
  }

  enqueue(op: BatchOp): void {
    this.queue.push(op);
    if (this.queue.length >= this.batchSize) {
      this.flush().catch((err) =>
        console.error("[BatchWriter] auto-flush error:", err)
      );
    }
  }

  async flush(): Promise<BatchResult> {
    if (this.queue.length === 0) {
      return { successful: 0, failed: 0, errors: [] };
    }

    const batch = this.queue.splice(0);
    const texts = batch.map(getEmbedText);
    const embeddings = await Embedder.getEmbeddings(texts);

    const bulkOps = batch.map((op, i) => ({
      index: INDEX_MAP[op.type],
      id: getId(op),
      body: { ...op.data, embedding: embeddings[i] },
    }));

    return this.db.bulkIndex(bulkOps);
  }

  async shutdown(): Promise<void> {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
    if (this.queue.length > 0) {
      await this.flush();
    }
  }
}

File: mairu/internal/batchwriter/batchwriter.go
package batchwriter

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type BatchOpType string

const (
	MemoryOp BatchOpType = "memory"
	SkillOp  BatchOpType = "skill"
	NodeOp   BatchOpType = "node"
)

type BatchOp struct {
	Type BatchOpType
	Data map[string]any
}

type BatchResult struct {
	Successful int
	Failed     int
	Errors     []map[string]string
}

type Embedder interface {
	GetEmbeddings(ctx context.Context, texts []string) ([][]float32, error)
}

type BulkIndexer interface {
	BulkIndex(ctx context.Context, ops []BulkOp) (BatchResult, error)
}

type BulkOp struct {
	Index string
	ID    string
	Body  map[string]any
}

type Options struct {
	BatchSize       int
	FlushIntervalMs int
}

var indexByType = map[BatchOpType]string{
	MemoryOp: "contextfs_memories",
	SkillOp:  "contextfs_skills",
	NodeOp:   "contextfs_context_nodes",
}

type Writer struct {
	embedder   Embedder
	indexer    BulkIndexer
	batchSize  int
	flushEvery time.Duration

	mu    sync.Mutex
	queue []BatchOp
	timer *time.Timer
}

func New(embedder Embedder, indexer BulkIndexer, opts Options) *Writer {
	batchSize := opts.BatchSize
	if batchSize <= 0 {
		batchSize = 10
	}
	flushMs := opts.FlushIntervalMs
	if flushMs <= 0 {
		flushMs = 2000
	}
	return &Writer{
		embedder:   embedder,
		indexer:    indexer,
		batchSize:  batchSize,
		flushEvery: time.Duration(flushMs) * time.Millisecond,
	}
}

func (w *Writer) Enqueue(op BatchOp) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.queue = append(w.queue, op)
	if len(w.queue) >= w.batchSize {
		if w.timer != nil {
			w.timer.Stop()
			w.timer = nil
		}
		return
	}
	if w.timer == nil {
		w.timer = time.AfterFunc(w.flushEvery, func() {
			_, _ = w.Flush(context.Background())
		})
	}
}

func embedText(op BatchOp) string {
	switch op.Type {
	case MemoryOp:
		if s, _ := op.Data["content"].(string); s != "" {
			return s
		}
	case SkillOp:
		return fmt.Sprintf("%v: %v", op.Data["name"], op.Data["description"])
	case NodeOp:
		return fmt.Sprintf("%v: %v", op.Data["name"], op.Data["abstract"])
	}
	return ""
}

func opID(op BatchOp) string {
	if op.Type == NodeOp {
		if uri, _ := op.Data["uri"].(string); uri != "" {
			return uri
		}
	}
	if id, _ := op.Data["id"].(string); id != "" {
		return id
	}
	return ""
}

func (w *Writer) Flush(ctx context.Context) (BatchResult, error) {
	w.mu.Lock()
	if w.timer != nil {
		w.timer.Stop()
		w.timer = nil
	}
	if len(w.queue) == 0 {
		w.mu.Unlock()
		return BatchResult{}, nil
	}
	batch := append([]BatchOp(nil), w.queue...)
	w.queue = nil
	w.mu.Unlock()

	texts := make([]string, 0, len(batch))
	for _, op := range batch {
		texts = append(texts, embedText(op))
	}
	embeddings, err := w.embedder.GetEmbeddings(ctx, texts)
	if err != nil {
		return BatchResult{}, err
	}

	ops := make([]BulkOp, 0, len(batch))
	for i, op := range batch {
		body := map[string]any{}
		for k, v := range op.Data {
			body[k] = v
		}
		if i < len(embeddings) {
			body["embedding"] = embeddings[i]
		}
		ops = append(ops, BulkOp{
			Index: indexByType[op.Type],
			ID:    opID(op),
			Body:  body,
		})
	}
	return w.indexer.BulkIndex(ctx, ops)
}
